File: packages/faasit-runtime/javascript/src/LocalRuntime.ts
import axios from "axios";
import { BaseFaasitRuntime, CallResult } from "./FaasitRuntime";

export class LocalRuntime extends BaseFaasitRuntime {

    name: string = "local"

    private event: any;
    constructor(event: any) {
        super()
        this.event = event;
    }
    async call(fnName: string, fnParams: {
        sequence?: number;
        input: object;
    }): Promise<CallResult> {
        console.log("call funtion");
        const axiosInstance = axios.create();
        const url = `http://master:9000/${fnName}`
        const resp = await axiosInstance.post(url, fnParams.input)
        console.log(resp.data);
        return { output: resp.data };
    }

    input() {
        return this.event;
    }

    output(returnObject: any) {
        return returnObject
    }
}

File: packages/faasit-runtime/python/faasit_runtime/runtime/local_runtime.py
from faasit_runtime.runtime.faasit_runtime import (
    FaasitRuntime, 
    InputType,
    CallResult,
    StorageMethods,
    FaasitRuntimeMetadata
)
import requests
import redis

class LocalRuntime(FaasitRuntime):
    name: str = 'local'
    def __init__(self, data, metadata:FaasitRuntimeMetadata) -> None:
        super().__init__()
        self._input = data
        self._storage = self.LocalStorage()
        self._metadata = metadata

    def input(self):
        return self._input

    def output(self, _out):
        return _out

    def metadata(self):
        return self._metadata

    async def call(self, fnName: str, fnParams: InputType) -> CallResult:
        resp = requests.post(
            f"http://{fnName}:9000", 
            json=fnParams, 
            headers={'Content-Type': 'application/json'}
        )
        result = resp.json()
        return result

    async def tell(self, fnName: str, fnParams: InputType):
        # 异步发送一个HTTP请求
        requests.post(
            f"http://{fnName}:9000", 
            json=fnParams, 
            headers={'Content-Type': 'application/json'}
        )
        pass

    @property
    def storage(self) -> StorageMethods:
        return self._storage

    class LocalStorage(StorageMethods):
        def __init__(self) -> None:
            self.redis_client = redis.Redis(host='redis', port=6379, db=0)
            pass

        async def set(self, key: str, value: str) -> None:
            self.redis_client.set(key, value)
            pass

        async def get(self, key: str) -> str:
            return self.redis_client.get(key)

        async def delete(self, key: str) -> None:
            self.redis_client.delete(key)
            pass

        async def list(self) -> list:
            return [key.decode('utf-8') for key in self.redis_client.keys()]

        async def exists(self, key: str) -> bool:
            return self.redis_client.exists(key)


        